File: scene3.c

#include <render_oo.h>
#include <scenedef.h>
#include <models.h>
#include <stdlib.h>
#include <time.h>
#include <render_ray.h>

float randf() {
    return (float) rand() / (uint64_t)RAND_MAX;
}

// scene.background = (.8, .8, .7)
// TODO: make scene.background work (I don't have image->clear)
int main() {
    init_scene();
    srand(time(0));
    struct Camera* cam = get_camera();
    struct Scene* scene = get_scene();
    cam->set_perspective(cam, 30, 1.3333, 5);

    struct Box b1 = make_box((struct Point3F32){-3, -2, -20}, (struct Vec3){2, 2, 2}, make_pixel(.8, 0, 0));
    struct Sphere s1 = make_sphere((struct Point3F32){2.5, -2, -20}, 1, make_pixel(0, .8, 0), 7, 15);
    struct Sphere s2 = make_sphere((struct Point3F32){0, 0, -25}, 3, make_pixel(randf(), randf(), randf()), 7, 15);
    struct Box b2 = make_box((struct Point3F32){0, -3.5, -20}, (struct Vec3){18, .5, 30}, make_pixel(.9, .8, .3));

    scene->add(scene, &(struct Shape){BOX, .shape.b = b1});
    scene->add(scene, &(struct Shape){SPHERE, .shape.s = s1});
    scene->add(scene, &(struct Shape){SPHERE, .shape.s = s2});
    scene->add(scene, &(struct Shape){BOX, .shape.b = b2});

    struct Image img_wf = make_image(640, 480);
    render_wireframe(scene, &img_wf);
    img_wf.save(&img_wf, "ppm/scene3-wf.ppm");

    struct Image img_rt = make_image(640, 480);
    raytrace(scene, &img_rt);
    img_rt.save(&img_rt, "ppm/scene3-rt.ppm");

    return 0;
}

File: examples/scene3.zig
const std = @import("std");

const groove_gfx = @import("groove_gfx");
const gfx_types = groove_gfx.gfx_types;
const image = groove_gfx.image;
const materials = groove_gfx.materials;
const models = groove_gfx.models;
const render_ray = groove_gfx.render_ray;
const scenedef = groove_gfx.scenedef;

pub fn main() !void {
    var debug_alloc = std.heap.DebugAllocator(.{}).init;
    const alloc = debug_alloc.allocator();

    try scenedef.init_scene(alloc);
    defer scenedef.deinit_scene(alloc);

    var cam = scenedef.get_camera();
    const scene = scenedef.get_scene();
    cam.set_perspective(30, 1.3333, 5);
    scene.background = .{ .r = 0.8, .g = 0.8, .b = 0.7 };

    var prng = std.Random.DefaultPrng.init(blk: {
        var seed: u64 = undefined;
        try std.posix.getrandom(std.mem.asBytes(&seed));
        break :blk seed;
    });

    var b1 = models.Shape{
        .box = models.Box.init(.{ .vals = .{ -3, -2, -20 } }, .{ .vals = .{ 2, 2, 2 } }, materials.make_material(gfx_types.Pixel{ .r = 0.8, .g = 0, .b = 0 })),
    };
    var s1 = models.Shape{
        .sphere = try models.Sphere.init(alloc, .{ .vals = .{ 2.5, -2, -20 } }, 1, materials.make_material(gfx_types.Pixel{ .r = 0, .g = 0.8, .b = 0 }), 20, 20),
    };
    const rand_r = prng.random().float(f32);
    const rand_g = prng.random().float(f32);
    const rand_b = prng.random().float(f32);
    var s2 = models.Shape{
        .sphere = try models.Sphere.init(alloc, .{ .vals = .{ 0, 0, -25 } }, 3, materials.make_material(gfx_types.Pixel{ .r = rand_r, .g = rand_g, .b = rand_b }), 20, 20),
    };
    var b2 = models.Shape{
        .box = models.Box.init(.{ .vals = .{ 0, -3.5, -20 } }, .{ .vals = .{ 18, 0.5, 30 } }, materials.make_material(gfx_types.Pixel{ .r = 0.9, .g = 0.8, .b = 0.3 })),
    };

    try scene.add(&b1);
    try scene.add(&s1);
    try scene.add(&s2);
    try scene.add(&b2);

    const cwd = try std.fs.cwd().realpathAlloc(alloc, ".");
    defer alloc.free(cwd);
    const img_path = try std.fs.path.join(alloc, &.{ cwd, "ppm/scene3-rt.ppm" });
    defer alloc.free(img_path);

    var img = try image.Image.init(alloc, 640, 480);
    defer img.deinit(alloc);
    render_ray.raytrace(scene, &img);
    try img.save(img_path);
}
